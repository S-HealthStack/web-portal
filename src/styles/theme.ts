import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
