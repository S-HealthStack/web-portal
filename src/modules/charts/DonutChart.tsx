import React, { useEffect, useRef, useState, useCallback, useMemo } from 'react';
import styled, { useTheme } from 'styled-components';
import * as d3 from 'd3';

import Tooltip from 'src/common/components/Tooltip';
import { SpecColorType } from 'src/styles/theme';
import { typography, px, colors } from 'src/styles';
import { drawPieChartShape } from './pieChartShape';
import { getTooltipContent } from './PieChart';
import { NO_RESPONSES_LABEL, getEmptyStateData } from './common-helpers';

const SVG_VIEWPORT_SIZE = 100;
const INNER_RADIUS_RATIO = (176 - 60) / 176; // where 176 is radius and 60 is sector height in reference design
const START_ANGLE = 0;

interface ContainerProps {
  $width: number;
  $height: number;
}

const Container = styled.div.attrs<ContainerProps>(({ $width, $height }) => ({
  style: {
    width: px($width),
    height: px($height),
  },
}))<ContainerProps>`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const SvgContainer = styled.svg`
  position: absolute;
  width: 100%;
  height: 100%;
`;

const LabelContainer = styled.div``;

const ValueLabel = styled.span<{ isEmptyState: boolean }>`
  ${({ isEmptyState }) =>
    isEmptyState ? typography.headingXMediumRegular : typography.headingXLargeSemibold};
  color: ${({ theme }) => theme.colors.textPrimary};
`;

const PercentageLabel = styled.span`
  ${typography.headingLargeSemibold};
  color: ${colors.textPrimary};
`;

type DonutChartDataItem = {
  value: number;
  color: SpecColorType;
  name: string;
  count: number;
  total: number;
};

type Props = {
  data: DonutChartDataItem[];
  totalPercents: number;
  width: number;
  height: number;
};

const getShapeId = (index: number) => `donut-chart-shape-${index}`;

function drawPieChartLabels(
  { container, pieData, arcGenerator }: ReturnType<typeof drawPieChartShape>,
  onMouseEnter: (event: React.MouseEvent<SVGPathElement, MouseEvent>, index: number) => void,
  onMouseLeave: (event: React.MouseEvent<SVGPathElement, MouseEvent>) => void
) {
  container
    .selectAll(null)
    .data(pieData)
    .enter()
    .append('g')
    .attr('transform', (d) => `translate(${pieData.length > 1 ? arcGenerator.centroid(d) : 0})`)
    .append('text')
    .text('')
    .attr('id', (d) => getShapeId(d.index))
    .attr('data-testid', (d) => getShapeId(d.index))
    .on('mouseenter', (event, d) => onMouseEnter(event, d.index))
    .on('mouseleave', onMouseLeave);
}

const TOOLTIP_THRESHOLD = 100;

const DonutChart = ({ data, totalPercents, width, height }: Props) => {
  const svgRef = useRef<SVGSVGElement>(null);
  const theme = useTheme();
  const [currentHoveredShape, setCurrentHoveredShape] = useState<number>(-1);

  const outerRadius = SVG_VIEWPORT_SIZE / 2;
  const innerRadius = outerRadius * INNER_RADIUS_RATIO;

  const isEmptyState = useMemo(() => !data.length, [data.length]);

  const filteredData = useMemo(() => data.filter((d) => d.value !== 0), [data]);

  const tooltipContent = useMemo(() => {
    const hoveredShapeData = filteredData[currentHoveredShape];
    if (!hoveredShapeData) {
      return undefined;
    }

    return getTooltipContent(hoveredShapeData);
  }, [currentHoveredShape, filteredData]);

  const tooltipContainer = useMemo(
    () =>
      currentHoveredShape < 0
        ? undefined
        : (d3
            .select(svgRef.current)
            .select(`#${getShapeId(currentHoveredShape)}`)
            .node() as HTMLElement),
    [currentHoveredShape]
  );

  const tooltipPosition = useMemo(() => {
    if (!svgRef.current || !tooltipContainer) {
      return 'r';
    }

    const labelRect = tooltipContainer?.getBoundingClientRect();
    const svgRect = svgRef.current.getBoundingClientRect();

    return svgRect.right - labelRect.right < TOOLTIP_THRESHOLD ? 'l' : 'r';
  }, [tooltipContainer]);

  const onLabelMouseEnter = useCallback(
    (_: React.MouseEvent<SVGPathElement, MouseEvent>, index: number) => {
      setCurrentHoveredShape(index);
    },
    []
  );

  const onLabelMouseLeave = useCallback(() => {
    setCurrentHoveredShape(-1);
  }, []);

  useEffect(() => {
    if (svgRef.current) {
      const pieChartShape = drawPieChartShape(
        {
          svg: svgRef.current,
          outerRadius,
          innerRadius,
          startAngleDegrees: START_ANGLE,
          data: isEmptyState
            ? [getEmptyStateData(theme.colors.background)]
            : filteredData.map(({ color, ...restData }) => ({
                ...restData,
                color: theme.colors[color],
              })),
        },
        onLabelMouseEnter,
        onLabelMouseLeave
      );

      drawPieChartLabels(pieChartShape, onLabelMouseEnter, onLabelMouseLeave);
    }
  }, [
    filteredData,
    innerRadius,
    isEmptyState,
    onLabelMouseEnter,
    onLabelMouseLeave,
    outerRadius,
    theme,
  ]);

  return (
    <Container $width={width} $height={height}>
      <SvgContainer ref={svgRef} viewBox={`0 0 ${SVG_VIEWPORT_SIZE} ${SVG_VIEWPORT_SIZE}`} />
      <LabelContainer>
        <ValueLabel isEmptyState={isEmptyState}>
          {isEmptyState ? NO_RESPONSES_LABEL : totalPercents}
        </ValueLabel>
        {!isEmptyState && <PercentageLabel>%</PercentageLabel>}
      </LabelContainer>
      <Tooltip
        static
        content={tooltipContent}
        container={tooltipContainer}
        show={!!tooltipContent && !!tooltipContainer}
        position={tooltipPosition}
        arrow
      />
    </Container>
  );
};

export default DonutChart;
