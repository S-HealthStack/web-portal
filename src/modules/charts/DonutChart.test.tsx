import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { act, findByTestId, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components/';
import { userEvent } from '@storybook/testing-library';
import DonutChart from 'src/modules/charts/DonutChart';
import theme from '../../styles/theme';

describe('DonutChart', () => {
  it('should render', async () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <DonutChart
          totalPercents={0}
          data={[
            {
              value: 10,
              total: 100,
              count: 100,
              color: 'primary',
              name: 'test',
            },
          ]}
          width={1000}
          height={1000}
        />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const label = await findByTestId(baseElement, 'donut-chart-shape-0');
    expect(label).toBeInTheDocument();

    act(() => {
      userEvent.hover(label);
    });

    act(() => {
      userEvent.unhover(label);
    });
  });

  it('[NEGATIVE] should render with empty data', () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <DonutChart totalPercents={0} data={[]} width={1000} height={1000} />
      </ThemeProvider>
    );

    expect(baseElement).toBeDefined();
  });
});
