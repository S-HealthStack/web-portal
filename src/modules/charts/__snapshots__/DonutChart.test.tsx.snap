// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DonutChart should render 1`] = `
.c4 {
  display: inline-block;
  position: relative;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: absolute;
  width: 100%;
  height: 100%;
}

.c2 {
  font-family: Inter,sans-serif;
  font-size: 3rem;
  line-height: 3.625rem;
  font-weight: 600;
  font-style: normal;
  color: #474747;
}

.c3 {
  font-family: Inter,sans-serif;
  font-size: 2rem;
  line-height: 2.625rem;
  font-weight: 600;
  font-style: normal;
  color: #474747;
}

<body>
  <div>
    <div
      class="c0"
      style="width: 62.5rem; height: 62.5rem;"
    >
      <svg
        class="c1"
        viewBox="0 0 100 100"
      >
        <g
          transform="translate(50, 50)"
        >
          <path
            d="M3.061616997868383e-15,-50A50,50,0,1,1,-3.061616997868383e-15,50A50,50,0,1,1,3.061616997868383e-15,-50M-6.053651791239757e-15,-32.95454545454545A32.95454545454545,32.95454545454545,0,1,0,6.053651791239757e-15,32.95454545454545A32.95454545454545,32.95454545454545,0,1,0,-6.053651791239757e-15,-32.95454545454545Z"
            data-testid="arc-shape-0"
            fill="#4475E3"
          />
          <g
            transform="translate(0)"
          >
            <text
              data-testid="donut-chart-shape-0"
              id="donut-chart-shape-0"
            />
          </g>
        </g>
      </svg>
      <div
        class=""
      >
        <span
          class="c2"
        >
          0
        </span>
        <span
          class="c3"
        >
          %
        </span>
      </div>
      <span
        class="c4"
        data-testid="tooltip-container"
      />
    </div>
  </div>
</body>
`;
